DeepSpeech.py

```diff
@@ -1710,56 +706,340 @@ def export():
+                                   FLAGS.task_index,
+                                   n_input,
+                                   n_context,
+                                   alphabet,
+                                   FLAGS.threads_per_set,
+                                   FLAGS.loader_buffer,
+                                   min(memory_limits),
+                                   FLAGS.queue_capacity,
+                                   allocate_indices=cluster.allocate_indices)
+        # react on data set (re-)starting (cluster wide event)
+        cluster.on_start_data_set = \
+            lambda data_set_index: model_feeder.start_data_set(data_sets[data_set_index])
+        # get all relevant graph end-points
+        optimizer, results_tuple, gradients, sample_number, mean_edit_distance, loss = \
+            get_tower_results(model_feeder)
+        # add summaries of all variables and gradients to log
+        log_grads_and_vars(gradients)
+        # op to merge all summaries for the summary hook
+        merge_all_summaries_op = tf.summary.merge_all()
+        # apply gradients to modify the model
+        apply_gradient_op = optimizer.apply_gradients(gradients)
+        # increment global sample counter
+        sample_counter = model_variable('sample_counter', None, 0, trainable=False)
+        sample_inc_op = tf.assign_add(sample_counter, sample_number)
+
+        # checkpoint manager to deal with all saving, restoring and result logging
+        checkpoint_manager = CheckpointManager(checkpoint_dir=FLAGS.checkpoint_dir,
+                                               load=FLAGS.load,
+                                               inter_secs=FLAGS.inter_secs,
+                                               keep_n_inters=FLAGS.keep_n_inters,
+                                               keep_n_epochs=FLAGS.keep_n_epochs)
+
+        # start feeding - feeding runs on every node
+        # Note: happens to result in very strange errors, if done earlier
+        # (supposedly during graph construction)
+        model_feeder.start_queue_threads(session, coord)
+
+        # in-graph replication approach
+        # all graph evaluation initiated by chief worker only
+        if is_chief:
+            # let the checkpoint manager either load a model or initialize all variables
+            epoch_history = checkpoint_manager.start(session)
+            # getting historic validation loss log for early stopping
+            dev_losses = [e[2] for e in epoch_history if not e[2] is None]
+            # retrieving number of applied samples from (potentially restored) model
+            n_samples_trained_on_model = session.run(sample_counter)
+            # number of samples per epoch - to be at least 1
+            n_samples_per_epoch = len(train_set.files)
+            assert n_samples_per_epoch > 0
+            # start epoch of our training
+            start_epoch = (n_samples_trained_on_model // n_samples_per_epoch) + 1
```

Comment:
Saving the number of trained samples in the model and using it with `len(train_set)` to calculate a start epoch breaks resuming with a different train set. Say you train for 10 epochs with a train set of 10 files. At the end of training, `n_samples_trained_on_model` = 100. Now you want to train for another 3 epochs on a set of 51 files. `n_samples_trained_on_model // len(train_set)` is 1, so it'll think it's starting from epoch 2, rather than epoch 11. `n_samples_already_trained` will be `100 % 51 = 49`, so it'll also skip 49 samples in the first epoch with the new data.

I think we should just use the epoch history returned by the checkpoint manager, and start from epoch `max(epoch_number, epoch_history) + 1`. Then the epoch count will be monotonically increasing, regardless of how many times you stopped and resumed, or if you changed the train set or not.